From: Fedor <fedork@ubuntu.(none)>
Date: Sat, 4 Sep 2010 01:14:23 -0400
Subject: [PATCH] [NET/IPv4/netfilter]: patches from upstream

 kernel.org commits:
 aeed9e82cd258b9699eaa6568efefba9cc6d5f01 [NET] loopback: Panic if registration fails
 9958da0501fced47c1ac5c5a3a7731c87e45472c net: remove time limit in process_backlog()
 b1e93a68ca41e7e73766f95ba32ca05cf9052e15 netfilter: conntrack: don't deliver events for racy packets

---
 drivers/net/loopback.c                    |    7 ++++++-
 include/net/netfilter/nf_conntrack_core.h |    3 ++-
 net/core/dev.c                            |    3 +--
 3 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/net/loopback.c b/drivers/net/loopback.c
--- a/drivers/net/loopback.c
+++ b/drivers/net/loopback.c
@@ -229,7 +229,12 @@ struct net_device loopback_dev = {
 /* Setup and register the loopback device. */
 static int __init loopback_init(void)
 {
-	return register_netdev(&loopback_dev);
+	int err = register_netdev(&loopback_dev);
+
+	if (err)
+		panic("loopback: Failed to register netdevice: %d\n", err);
+
+	return err;
 };
 
 module_init(loopback_init);
diff --git a/include/net/netfilter/nf_conntrack_core.h b/include/net/netfilter/nf_conntrack_core.h
--- a/include/net/netfilter/nf_conntrack_core.h
+++ b/include/net/netfilter/nf_conntrack_core.h
@@ -67,7 +67,8 @@ static inline int nf_conntrack_confirm(struct sk_buff **pskb)
 	if (ct && ct != &nf_conntrack_untracked) {
 		if (!nf_ct_is_confirmed(ct))
 			ret = __nf_conntrack_confirm(pskb);
-		nf_ct_deliver_cached_events(ct);
+		if (likely(ret == NF_ACCEPT))
+			nf_ct_deliver_cached_events(ct);
 	}
 	return ret;
 }
diff --git a/net/core/dev.c b/net/core/dev.c
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -1950,7 +1950,6 @@ static int process_backlog(struct net_device *backlog_dev, int *budget)
 	int work = 0;
 	int quota = min(backlog_dev->quota, *budget);
 	struct softnet_data *queue = &__get_cpu_var(softnet_data);
-	unsigned long start_time = jiffies;
 
 	backlog_dev->weight = weight_p;
 	for (;;) {
@@ -1971,7 +1970,7 @@ static int process_backlog(struct net_device *backlog_dev, int *budget)
 
 		work++;
 
-		if (work >= quota || jiffies - start_time > 1)
+		if (work >= quota)
 			break;
 
 	}
-- 
